ts/utils/fetch.ts

```diff
@@ -114,3 +129,246 @@ export const constantPollingFetch = (
+ */
+export function lollipopFetch(
+  timeout: Millisecond = fetchTimeout,
+  maxRetries: number = fetchMaxRetries,
+  lollipopConfig: LollipopConfig,
+  keyInfo: KeyInfo
+) {
+  const fetchApi = (global as any).fetch;
+  const abortableFetch = AbortableFetch(fetchApi);
+  const timeoutFetch = toFetch(
+    setFetchTimeout(timeout, abortableFetch)
+  ) as FixedFetch;
+  // configure retry logic with default exponential backoff
+  // @see https://github.com/pagopa/io-ts-commons/blob/master/src/backoff.ts
+  const exponentialBackoff = calculateExponentialBackoffInterval();
+  const retryLogic = withRetries<Error, Response>(
+    maxRetries,
+    exponentialBackoff
+  );
+  const retryWithTransient429s = retryLogicForTransientResponseError(
+    _ => _.status === 429,
+    retryLogic
+  );
+
+  return retriableFetch(retryWithTransient429s)(
+    (input: RequestInfo | URL, init?: RequestInit) => {
+      const requestAndKeyInfo = isKeyInfoAndRquestValid(keyInfo, input, init);
+      if (requestAndKeyInfo) {
+        // eslint-disable-next-line functional/no-let
+        let newInit = requestAndKeyInfo.init;
+        const { body, bodyString, inputUrl, method, originalUrl } =
+          extractHttpRequestComponents(
+            requestAndKeyInfo.input,
+            requestAndKeyInfo.init
+          );
+        if (body) {
+          newInit = addHeader(
+            newInit,
+            "Content-Digest",
+            generateDigestHeader(bodyString)
+          );
+        }
+        const signatureConfigForgeInput: SignatureConfigForgeInput = {
+          publicKey: requestAndKeyInfo.publicKey,
+          keyTag: requestAndKeyInfo.keyTag,
+          lollipopConfig,
+          method,
+          inputUrl,
+          originalUrl
+        };
+        const mainSignatureConfig: SignatureConfig = forgeSignatureConfig(
+          signatureConfigForgeInput,
+          keyInfo,
+          [
+            "Content-Digest",
+            "Content-Type",
+            "x-pagopa-lollipop-original-method",
+            "x-pagopa-lollipop-original-url"
+          ]
+        );
+        newInit = addHeader(
+          newInit,
+          "x-pagopa-lollipop-original-method",
+          method
+        );
+        newInit = addHeader(
+          newInit,
+          "x-pagopa-lollipop-original-url",
+          originalUrl
+        );
+        const newInitHeaders =
+          (newInit.headers as Record<string, string>) ?? {};
+        const {
+          signatureBase: mainSignatureBase,
+          signatureInput: mainSignatureInput
+        } = generateSignatureBase(newInitHeaders, mainSignatureConfig, 1);
+        const customSignPromises: Array<Promise<SignPromiseResult>> =
+          lollipopConfig.customSignatures?.map((headerValue, index) => {
+            const headerIndex: number = index + 2;
+            const headerName = `x-pagopa-lollipop-custom-${headerIndex}`;
+            const customHeader = {
+              [headerName]: headerValue
+            };
```

Comment:
If I'm not mistaken, the headers containing the custom contents to be signed are added here. My suggestion was to let the single product (the function caller) build the headers as needed. In this way, the single `lollipoConsumer` (LC) has no knowledge of which header (therefore which `signature`) is associated with the value sent. (`customSignature`)

Reply:
@grausof is the actual implementation already suitable for FCI?

Reply:
that could be fine, but there must be a way to be able to associate the signature `(sig2, sig3...) `with the custom input (`customSignatures`), which could also be the index of the input array. But that doesn't seem like a clean solution to me

Reply:
The actual solution uses the array index as suffix `x-pagopa-lollipop-custom-${headerIndex}`, it's right.  One solution would be to change `customContentToSig` (`customSignatures`) from `Array<string>` to `Record<string, string>` where the **key/value** pairs are **headerNameSuffix/headerValue**. Ex: { "sig2" :  "ASDFFA324SDFA==",  "sig3": "DAFDEFAF323DSFA==" }. But I would use this "headerName" as suffix like that `x-pagopa-lollipop-custom-${headerNameSuffix}` so that to be sure those headers cannot collide with other default headers or any other future custom headers and to be sure to have the `x-` prefix as the preferred way to start a custom header name. What do you think?

Reply:
We have changed the type of custom content to sign from Array<string> to Record<string, string>